xtheadmaee.adoc

```diff
@@ -0,0 +1,34 @@
+[#xtheadmaee]
+== Extend MMU address attribute
+
+[NOTE,caption=Frozen]
+The `XTheadMaee` extension is `stable`.
+
+The `XTheadMaee` ISA extension provides a way to configure page attributes of addresses when virtual addresses are translated into physical addresses by extended page attributesin page table entries (PTEs).
+
+Extension version: 1.0.
+
+The table below describes the page attributes corresponding to each memory type:
+
+[cols="^3,^3,12,18",stripes=even,options="header"]
+|===
+| Memory type            | SO | C | B
+| Cacheable memory       | 0  | 1 | 1
+| Non-cacheable memory   | 0  | 0 | 1
+| Bufferable device      | 1  | 0 | 1
+| Non-bufferable device  | 1  | 0 | 0
+|===
+
+The table below describes extend page attributes in PTE:
+[cols="^3,^3,12",stripes=even,options="header"]
+|===
+| Fields  | Bit | Meaning
+| SO      | 63  | Strong order
+| C       | 62  | Cacheable
+| B       | 61  | Bufferable
+| T       | 60  | Trustable
+|===
+
+XTheadMaee depends on the value of the MAEE field in extended register MXSTATUS. If the MAEE field is 1, page attributes of addresses are determined by extended page attributes in corresponding PTEs. If the MAEE field is 0, page attributes of addresses are determined by the sysmap.h file.
+
```

Comment:
How can we find out if XTheadMaee is available on a system (how can the kernel find out if XTheadMaee is present or not).

Reply:
Usually it is the uboot spl or the zero stage boot, which run before the OpenSBI and the kernel, do the enable work by write 1 to TH_MXSTATUS MAEE field. Util now, the Linux kernel directly use the  XTheadMaee. It doesn't detect. I think we should do it in the OpenSBI and export an interface to Linux kernel if we don't have a supervisor register.

Reply:
Linux does not unconditionally write or read a bit in a custom CSR.
Instead, it needs to detect first that the kernel is running on a core that implements MAEE.
And that detection mechanism should be documented here (so the code can be reviewed against the specification).

Reply:
Yes, we can export the XTheadMaee through DTB. But util now, Linux can't read or write TH_MXSTATUS. It is not exposed in menvcfg like SVPBMT.

Reply:
If MVENDORID is 0x5B7, OS kernel should read TH_MXSTATUS MAEE as the enable status of XTheadMaee. Is this OK？

Reply:
The current kernel implementation can be found here:
* https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/riscv/kernel/alternative.c
* https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/riscv/errata/thead/errata.c.

The following requirements are checked:
* static test: `CONFIG_ERRATA_THEAD` (`riscv_fill_cpu_mfr_info`)
* `vendor_id == THEAD_VENDOR_ID` (`riscv_fill_cpu_mfr_info`)
* static test: `CONFIG_ERRATA_THEAD_PBMT` (`errata_probe_pbmt`)
* `arch_id == 0` (`errata_probe_pbmt`)
* `impid == 0` (`errata_probe_pbmt`)

If I understand you correct, then you state that this is not sufficient and future proof.
You would also like to see a test for a set `MAEE` bit in `TH_MXSTATUS` (in `errata_probe_pbmt`).
This looks fine to me.

But now we have the question of the requirements for accessing `TH_MXSTATUS.MAEE`.
Is this CSR available for all cores with `(vendor_id == THEAD_VENDOR_ID) && (arch_id == 0) && (impid == 0)` (follow-up question: what about future cores?). If this test is sufficient, then we are done. If not, we need to clarify what is needed.

Reply:
This CSR available for all cores in Xuantie CPU. One shortcoming using TH_MXSTATUS.MAEE is that it can't be read from supervisor mode．If you don't think it is a problem, I am happy to add TH_MXSTATUS.MAEE as XTheadMaee proof.